src/testModel.ts

```diff
@@ -709,10 +709,14 @@ export class TestModelCollection extends DisposableBase {
 
   clear() {
     this.dispose();
```

Comment:
I was suggesting that this is the actual problem, i.e. no method in the class should be calling `dispose()` since after calling dispose, we can't use that instance. Imagine one of the disposables is essential to the instance operation, now clearing its state will be corrupting the instance.

Reply:
I understand, `dispose` should call `clear`, not this way around. I think that extending `TestModelCollection` extending `DisposableBase` in this case doesn't help, because it implicitly implements the `dispose` method the wrong way, and to make it right we need to call `super.dispose()` in `clear` and then `this.clear()` in the overriden `dispose` method.

Reply:
Actually, dispose cannot call `this.clear()` because it would fire `onUdate` event. That's why I added the `_disposeModels` function in one of my attempts, which was then called in both `clear` and `dispose` functions